### ISSUE TITLE
Handling paths as nicknames

### ISSUE THREAD

**Issue by lukechampine** [NebulousLabs/Sia#894]
The new API treats nicknames as paths. Even `/renter/upload/foo.txt` will create a file called `/foo.txt`, with a leading slash. What is the right way to handle these strings?

There are three areas that need attention:
- The API. I see no problems here (for now), but we might consider dropping the leading slash from nicknames.
- The renter. Currently the renter tries to save .sia files using their nickname as the filename, so with the new API it will be creating files with slashes in their names, which is a disaster. We probably want to perform some kind of escaping here. Alternatively, the .sia filenames don't *really* need to match their nicknames; they could be numbers (like the host) or something else entirely.
- The UI+siac. Things are pretty clean for the UI, since path handling happens behind the scenes, but in siac the nicknames are more in-your-face. For example, you'll need to include the leading slash with some operations, but not others (namely `rename`, since it takes a path as a URL parameter). We could play fast and loose here (tacking on leading slashes behind the scenes), or we could be strict and require an explicit leading slash in all calls.

A more radical alternative would be to ban `/` in nicknames altogether. This is basically taking the position that siad should not know anything about paths, and that only the UI should be responsible for manipulating/parsing paths, using some other special character in place of `/`. I can sympathize with this position, but I feel it will become increasing untenable as the software matures.

I plan to implement this stuff myself, but I'm looking for input as to what seems like a good design.

**Comment by Mingling94** [comment]
* I think file nicknames/paths should drop the leading slash and enable freedom for nickname/paths. As I mentioned to you in slack, uploading a file with `/renter/upload/foo` with make it impossible to delete 

* Is it hard to allow a path/nickname parameter in the url to `/renter/load` and `/renter/loadascii` like for the other endpoints? This would seem the most congruent.

* Tack on the leading slashes behind the scenes IMO. siac's a UI just like the GUI and can aim to be more convenient for a user. Most CLI tools I can think of allow file paths to start with a file or folder name and do not require a `./`

**Comment by DavidVorick** [comment]
Allowing slashes in filenames is very convenient, and I think the right way to do it. Forcing things to use another character strikes me as completely awful, and forcing the UI to handle the pathing also strikes me as awful.

I don't really know what to do about the leading slash. Half of me thinks that the leading slash makes sense, but the other half of me really doesn't. It should be pretty easy to decide to always include it or never include it though, right? And, if we're following a unix philosophy it makes sense to have the leading slash, since that's how files in unix are set up. It helps you know whether you are talking about a relative path vs. an absolute path.

As for the renter storing things on disk, it should also be pretty easy to remove the leading slash if you can always expect it to be there (after being processed so that it's changed from a relative path to an absolute path, but the api only deals with absolute paths). I don't know that it makes sense for the renter to have files on disk that share the same name as their nickname. On one hand, it makes it really easy to find `.sia` files if you are looking to import or export a whole cluster of them. On the other, it means that the renter is going to need to have a directory structure.

And, it confuses things when you start exporting. When you export, do you leave the folder or only take the trailing name? Intuition tells me that you should only take the trailing name, unless the `.sia` file is an entire folder, in which case you should take the trailing name for the highest level directory that's being exported.

In summary:

- keep the leading slash
- allow slashes in the nicknames via api
- have the renter save the files with names that aren't correlated to their actual names. I don't really like this b/c it means people can't go in and copy-paste files, but I'd probably rather they use the tools anyway. The only other option would be to recreate the folder structure within the renter, but I don't think I like that

**Comment by Mingling94** [comment]
@DavidVorick I don't understand your last bullet point. Does that only apply to .sia files or what?

**Comment by DavidVorick** [comment]
@Mingling94 The last point pertains only to the renter backend, and not to the api or any of the frontends.

After further thought (and minimal discussion with Luke), I'm now favoring the renter backend keeping a full directory structure, which also makes it possible to have empty folders. We should also consider storing the file permissions, including a flag indicating whether a file is a directory.

Luke has convinced me that a leading slash is not needed, and that we should drop it, and assume that all paths going into and coming out of the api are absolute paths. This seems okay to me.

**Comment by Mingling94** [comment]
So should I expect a PR removing the need for the '/' soon?

**Comment by lukechampine** [comment]
I'm sure we'll see a Sia filesystem via FUSE at some point, but it may not be developed internally :P

My thoughts:
- Leading slash should definitely be dropped; I think it's fair to argue that all Sia paths are already "relative" in regards to the actual filesystem they're being used on.
- When you create a .sia file for `foo/bar/baz,jpg`, you're probably not intending to recreate that folder structure when the .sia file is downloaded (you just want `baz.jpg`). So it stands to reason that only the base of the path should be used. To share multiple files, you'd have to create a folder for them and then share the folder, e.g. `/renter/share/foo/bar` would create a .sia file containing all the files in `bar`, and would cause the `bar` folder structure to be replicated upon download. This has the benefit of letting us return to the `/renter/share/<path>` route structure for sharing, instead of a URL parameter containing comma-separated nicknames. The downside is that you are forced to create a folder to share multiple files, but IIRC BitTorrent has this requirement as well.
- Full directory structure in the `renter` folder seems more and more sane to me. It removes the need for any sort of special nickname parsing/escaping. In a way, it's "outsourcing" the problem of directory structure to the operating system, which intuitively feels like a smart thing to do. Essentially, we can use all the standard tools to manipulate this "meta-directory" as a means of manipulating the structure of the data it actually points to.

If there are no objections, I'm going to implement 1) right away, 3) soon after (since it's currently completely broken), and 2) if I get around to it later, since it's non-critical.

**Comment by Mingling94** [comment]
:+1: to everything luke said. Seems best in all regards.

**Comment by DavidVorick** [comment]
:+1:
